validator/testcases/frameworks.py

```diff
@@ -0,0 +1,546 @@
+import json
+import re
+
+from validator.constants import LIBRARY_METADATA
+from validator.decorator import cached_property
+from validator.testcases import libraries
+from validator.version import Version
+
+
+def fake_version(files):
+    """
+    Returns a hash to act in place of a version number, given the contents
+    of a set of known files.
+    """
+    import hashlib
+    return hashlib.sha256("".join(files)).hexdigest()[:16]
+
+
+class MetaFramework(type):
+    def __new__(mcls, name, bases, dict_):
+        cls = super(MetaFramework, mcls).__new__(mcls, name, bases, dict_)
+        if cls.NAME:
+            cls.FRAMEWORKS.append(cls)
+        return cls
+
+
+class Framework(object):
+    """
+    Base class for framework definitions.
+    """
+
+    FRAMEWORKS = []
+
+    SAFE_FILES = ".jpg", ".ico", ".png", ".gif", ".txt"
+
+    CHECKED_FILES = re.compile(r"\.(jsm?|xul|xbl|x?html?)$", re.I)
+
+    __metaclass__ = MetaFramework
+
+    def __init__(self, err, xpi):
+        self.err = err
+        self.xpi = xpi
+
+        # Update certain metadata from the metadata JSON.
+        # Ideally this would also update based on data for the
+        # specific version of the framework, but there are
+        # bootstrapping issues.
+        for prop in "framework_files", "framework_prefixes":
+            if prop in self.framework_metadata:
+                setattr(self, prop.upper(),
+                        tuple(self.framework_metadata[prop]))
+
+    NAME = None
+    ACCEPTED = True
+
+    # Path of the JSON file within the XPI containing framework
+    # metadata. Not applicable to all frameworks.
+    METADATA_FILE = None
+
+    # Tuple of file paths which, if they exist, belong to the framework.
+    FRAMEWORK_FILES = ()
+
+    # Tuple of file path prefixes for which matching files, if they
+    # exist, belong to the framework.
+    FRAMEWORK_PREFIXES = ()
+
+    @cached_property
+    def ID(self):
+        """Internal ID of the framework."""
+        return self.NAME and self.NAME.lower()
```

Comment:
This will return `False` if there's no `self.NAME`, and not an empty string or `None`. I understand this is only missing for the base `Framework` class, so maybe you don't need this test?